torch/csrc/autograd/profiler.cpp

```diff
@@ -418,9 +504,67 @@ void Event::record(bool record_cuda) {
   cpu_ns_ = getTime();
 }
 
-double Event::cuda_elapsed_us(const Event & e) {
+/* static */ Event Event::fromIValue(const at::IValue& eventIValue) {
+  TORCH_INTERNAL_ASSERT(
+      eventIValue.isList(),
+      "Expected IValue to contain type c10::impl::GenericList");
+  auto ivalues = eventIValue.toList();
+  TORCH_INTERNAL_ASSERT(
+      ivalues.size() >= kEventIValuesSize,
+      "Expected at least ",
+      kEventIValuesSize,
+      " elements to reconstruct Event.");
+
+  Event evt(
+      static_cast<EventKind>(
+          ivalues.get(EventIValueIdx::KIND).toInt()), // EventKind
+      at::StringView(ivalues.get(EventIValueIdx::NAME).toStringRef()), // name
+      ivalues.get(EventIValueIdx::THREAD_ID).toInt(), // thread_id
+      static_cast<at::RecordFunctionHandle>(
+          ivalues.get(EventIValueIdx::HANDLE).toDouble()), // handle
+      {}, // TODO: record shapes
+      ivalues.get(EventIValueIdx::NODE_ID).toInt(), // node id
+      true, // is remote
+      ivalues.get(EventIValueIdx::CPU_MEM_USAGE).toInt(), // cpu_mem_usage
+      ivalues.get(EventIValueIdx::CPU_NS).toInt(), // cpu_ns
+      ivalues.get(EventIValueIdx::CUDA_RECORDED).toBool(), // was cuda recorded
+      ivalues.get(EventIValueIdx::CUDA_MEM_USAGE).toInt(), // cuda memory usage
+      ivalues.get(EventIValueIdx::CUDA_DEVICE).toInt(), // device
+      ivalues.get(EventIValueIdx::CUDA_US).toInt() // cuda_us
+  );
+  return evt;
+}
+
+at::IValue Event::toIValue() const {
+  c10::impl::GenericList eventIValueList(at::AnyType::get());
+  eventIValueList.reserve(kEventIValuesSize);
+  eventIValueList.emplace_back(static_cast<int64_t>(kind_));
+  eventIValueList.emplace_back(std::string(name_.str()));
+  eventIValueList.emplace_back(thread_id_);
+  eventIValueList.emplace_back(static_cast<double>(handle_));
+  eventIValueList.emplace_back(node_id_);
+  eventIValueList.emplace_back(cpu_memory_usage_);
+  eventIValueList.emplace_back(cpu_ns_);
+  // CUDA event information
+  bool cuda_profiling_enabled = has_cuda();
+  eventIValueList.emplace_back(cuda_profiling_enabled);
+  eventIValueList.emplace_back(static_cast<int64_t>(cuda_memory_usage_));
+  eventIValueList.emplace_back(device_);
+  eventIValueList.emplace_back(cuda_us_);
+  return at::IValue(eventIValueList);
+}
+
+double Event::cuda_elapsed_us(const Event & e) const {
   TORCH_CHECK(e.has_cuda() && has_cuda(), "Events were not recorded for CUDA");
-  TORCH_CHECK(e.device() == device(), "Events are not on the same device");
+  TORCH_CHECK(
+      e.device() == device(),
+      c10::str(
+          "Events are not on the same device: ", e.device(), " vs ", device()));
+  if (isRemote() && e.isRemote()) {
+    // validate that cuda_us_ has been set properly.
+    TORCH_INTERNAL_ASSERT(cuda_us_ >= 0 && e.cuda_us_ >=0);
+    return e.cuda_us_ - cuda_us_;
```

Comment:
@ilia-cher I refactored to change `cuda_us_` to `int64_t`,so  I think it should also be fine to change this function to return `int64_t` as well